internal/utils/denos/build.ts

```diff
@@ -72,3 +76,127 @@ async function buildAndWrite(p: string, importMapPath: string) {
 
 buildAndWrite(Deno.args[0], Deno.args[1]);
 
+// Adapted from https://github.com/denoland/deno/blob/bacbf949256e32ca84e7f11c0171db7d9a644b44/cli/auth_tokens.rs#L38
+
+function parseDenoAuthTokens(
+  denoAuthTokens: string | undefined,
+): [string, string][] {
+  const tokens: [string, string][] = [];
+
+  if (denoAuthTokens) {
+    denoAuthTokens.split(";").forEach((tokenAndHost) => {
+      if (!tokenAndHost.includes("@")) {
+        console.warn("Badly formed auth token discarded.");
+        return;
+      }
+
+      const sepIdx = tokenAndHost.lastIndexOf("@");
+      const token = tokenAndHost.substring(0, sepIdx);
+      const host = tokenAndHost.substring(sepIdx + 1);
+
+      if (token.includes(":")) {
+        const sepIdx = token.lastIndexOf(":");
+        const username = token.substring(0, sepIdx);
+        const password = token.substring(sepIdx + 1);
+        tokens.push([host, `Basic ${encode(`${username}:${password}`)}`]);
+      } else {
+        tokens.push([host, `Bearer ${token}`]);
+      }
+    });
+  }
+
+  return tokens;
+}
+
+function findMatchingDenoAuthToken(
+  hostToMatch: string,
+  denoAuthTokens: [string, string][],
+): string | undefined {
+  const matchingPair = denoAuthTokens.find(([host, _]) =>
+    hostToMatch.endsWith(host)
+  );
+  return matchingPair?.[1];
+}
+
+// Adapted from https://github.com/denoland/eszip/blob/b7043bf0b7938f8d91216e1541d4bd4fb8819a3d/lib/loader.ts
+
+interface LoadResponseModule {
+  /** A module with code has been loaded. */
+  kind: "module";
+  /** The string URL of the resource. If there were redirects, the final
+   * specifier should be set here, otherwise the requested specifier. */
+  specifier: string;
+  /** For remote resources, a record of headers should be set, where the key's
+   * have been normalized to be lower case values. */
+  headers?: Record<string, string>;
+  /** The string value of the loaded resources. */
+  content: string;
+}
+
+interface LoadResponseExternalBuiltIn {
+  /** The loaded module is either _external_ or _built-in_ to the runtime. */
+  kind: "external" | "builtIn";
+  /** The string URL of the resource. If there were redirects, the final
+   * specifier should be set here, otherwise the requested specifier. */
+  specifier: string;
+}
+
+type LoadResponse = LoadResponseModule | LoadResponseExternalBuiltIn;
+
+async function load(
```

Comment:
We should probably continue to use `load` function from eszip module in the long-term. Otherwise we'd miss any updates/security fixes. Don't mind we reimplement it for the short-term. 

Maybe we can propose a change to eszip package to accept a custom function that provides request headers? eg: `load(specifier: string, headerFunc: (url: URL) => Headers)`

Reply:
Yeah makes sense, I'll add a task to upstream this as a follow up